### DISCUSSION TITLE
How to temporarily stop the scroll (Modal case)

### DISCUSSION THREAD

**Discussion by HectorLS** [idiotWu/smooth-scrollbar#361] [FAQ] [answered]
"smooth-scrollbar": "^8.2.4" 

Again, awesome scroll 👍 

Hi, maybe is a silly question, but i didn't find a method to block the scroll on the page ( without destroy it )

like for example if

>  i have opened a modal box, and i want to temporarily disable the scroll on the page 💻 

Thanks in advance

**Comment by dolphin-wood** [comment, marked as answer]
You can use plugin system to do this:

```js
import Scrollbar, { ScrollbarPlugin } from 'smooth-scrollbar';

class ModalPlugin extends ScrollbarPlugin {
  static pluginName = 'modal';

  static defaultOptions = {
    open: false,
  };

  transformDelta(delta) {
    return this.options.open ? { x: 0, y: 0 } : delta;
  }
}

Scrollbar.use(ModalPlugin, /* OverscrollPlugin */);

// usage
const scrollbar = Scrollbar.init(elem);

onModalOpen  -> scrollbar.updatePluginOptions('modal', { open: true });
onModalClose -> scrollbar.updatePluginOptions('modal', { open: false });
```

**Reply by Minozzzi** [reply]
How I back to native scroll instead of stop completely scroll?

**Reply by dolphin-wood** [reply]
You can call `scrollbar.destroy()` to remove the smooth scrollbar completely.

**Reply by Minozzzi** [reply]
And then I need to build again with `Scrollbar.init`?

**Reply by dolphin-wood** [reply]
Correct. This is the only way to switch between the native scrollbars and ours.

**Reply by Minozzzi** [reply]
I understand, maybe there is another way to solve my problem?

The problem is the following, in a certain container I have an animation that is on top of the `scroll` and `wheel` event of the `window`. And when I have the scrollbar instance `Scrollbar.init`, the animation doesn't work. Basically in my animation I change the `scale` of an image according to `deltaY`. Is there another way to work my animation with the `scrollbar`?

**Reply by dolphin-wood** [reply]
I believe you can attach a listener to the scrolling event using:

```js
scrollbar.addListener(({ offset, limit }) => {
  // your animation...
});
```

Here is the doc: https://github.com/idiotWu/smooth-scrollbar/blob/develop/docs/api.md#scrollbaraddlistener

**Reply by dolphin-wood** [reply]
...and if you want to access the `deltaY`, you can create a plugin like:

```js
class MyPlugin extends ScrollbarPlugin {
  static pluginName = 'myPlugin';

  transformDelta(delta, fromEvent) {
    // do something with `delta.y`
    return delta;
  }
}
```

Docs: https://github.com/idiotWu/smooth-scrollbar/blob/develop/docs/plugin.md#transformdelta

**Reply by Minozzzi** [reply]
Thanks!! I will try to work in the two ways mentioned

**Reply by Minozzzi** [reply]
Hey bro, in my case I create the plugin, but I need access to refs and react states in which my component will have the deltaY logic. Is there a way to do this?
![image](https://user-images.githubusercontent.com/61752597/227350199-f626bfa6-f053-44c0-841f-f6118935eadd.png)

**Reply by dolphin-wood** [reply]
In that case, you can pass the `handleContainerScroll()` callback as an option for your plugin. For example:

```js
class NhShapeAnimationPlugin extends ScrollbarPlugin {
  static pluginName = 'NhShapeAnimationPlugin';

  static defaultOptions = {
    handleContainerScroll: null,
    isOnContainer: false,
  };

  transformDelta(delta, event) {
    if (/wheel/.test(event.type) && this.options.handleContainerScroll) {
      this.options.handleContainerScroll(event);
    }

    return this.options.isOnContainer ? { x: 0, y: 0 } : delta;
  }
}

// usage
function Component() {
  useEffect(() => {
    function handleContainerScroll(event) {
      // access refs and states here
    }

    // init the scrollbar
    Scrollbar.init(target, {
      plugins: {
        NhShapeAnimationPlugin: {
          handleContainerScroll,
        },
      },
    });

    // or update the plugin option
    scrollbar.updatePluginOptions('NhShapeAnimationPlugin', {
      handleContainerScroll,
    });
  }, []);
}
```

Alternatively, you can also pass the refs and states as the options and update them inside your components.

These solutions are a bit tricky and chaotic for now, but I'll be sure to refactor the plugin system in the next major release! See  https://github.com/idiotWu/smooth-scrollbar/issues/411

**Reply by Minozzzi** [reply]
This seems to work!!
I'll test it tomorrow, but in theory it seems to work. And you don't have to worry about complexity, this lib is the only one I thought actually did what it was intended to do, so I really appreciate the help.

I will try to dedicate some time to contribute to this amazing project

**Reply by Minozzzi** [reply]
Hey @idiotWu your solutions worked great. Thank you bro!!

**Comment by HectorLS** [comment]
Brilliant, way better than my approach ! 🎉 
Thank you

**Comment by vibesudi** [comment]
It's so cool, Thanks.

**Comment by sett-dr-cat911** [comment]
thank you

**Comment by sadeghbarati** [comment]
@idiotWu hi
its `block` scroll 
but when u **`hover the corner`** trackbar become visible and can drag it to scroll

how `remove` it while modal is open and how to `reEnable` it when modal is close

**Comment by robizzt** [comment]
I have same problem, where I use dropdown as modal on desktop resolutions.

I noticed if you create modal outside of main smooth scroll wrapper and initalize modal smooth scroll, you wont be able to scroll main scroll as long as mouse is over modal. So, if you have modal over whole screen (or modal backdrop), everything should work as expected, without disabling or destroying any smooth scrollbar.

So, I think best partice would be, if you have more than one smooth scrollbar wrappers, you should place each outside of it.
If you have one inside of other, after inner is scrolled to end, outter will start to scroll.

**Comment by sadeghbarati** [comment]
@robizzt 
yeah you right I check Cuberto website which uses Smoothscroll

`fixed menu` outside of main wrapper of smoothscrollbar when its was fixed, smoothscroll work pretty well and you can hide and show scrollbar track yAxis with z-index

but when you have `static menu` outside of wrapper of smoothscroll it gets messy and feels like it's fixed and buggy on scroll and you can drag trackbar and two scrollbar appears one native one smoothscroll


I test everything coz I needed to header and menu outside of wrapper and at last I find no solution for it
and move my menu inside of wrapper to work 

``` javascript
this.scrollbar.track.yAxis.element.style.display = "none"
```

**Comment by robizzt** [comment]
I decided to go with `.destroy()` because I have dropdown that behaves like a modal. When I open dropdown, I destroy main smoothscroll, when I close dropdown, I initialize main smoothscroll again. I don't know if that's bad practice, cause it is a little complicated, but it works for me.

**Reply by robizzt** [reply]
But with Plugin it's better and cleaner.

**Comment by toxifiedmatrix** [comment]
> You can use plugin system to do this:
> 
> ```js
> import Scrollbar, { ScrollbarPlugin } from 'smooth-scrollbar';
> 
> class ModalPlugin extends ScrollbarPlugin {
>   static pluginName = 'modal';
> 
>   static defaultOptions = {
>     open: false,
>   };
> 
>   transformDelta(delta) {
>     return this.options.open ? { x: 0, y: 0 } : delta;
>   }
> }
> 
> Scrollbar.use(ModalPlugin, /* OverscrollPlugin */);
> 
> // usage
> const scrollbar = Scrollbar.init(elem);
> 
> onModalOpen  -> scrollbar.updatePluginOptions('modal', { open: true });
> onModalClose -> scrollbar.updatePluginOptions('modal', { open: false });
> ```

How can I implement the same in JQuery, please help me.

**Reply by robizzt** [reply]
Just remove "import" line and change "ScrollbarPlugin" in _class ModalPlugin extends ScrollbarPlugin {_ to Scrollbar.ScrollbarPlugin. No need for jQuery.

**Comment by PavelZhuravlev** [comment]
@idiotWu , Thank you for amazing library!
I'm facing small issue with "Temporarily stop plugin" and can you please help me understand
Is it possible to temporary stop scrolling in some specific point?

For example:
I have scroll event listener and im trying to temporary block scrolling when offset become 500px from top.
But when i am trying to use plugin from this topic, time to time i'm facing overscrolls. Scrolling can stops after 550px, 600px, 700px, because of some overscrolling (remainMomentum) + addListener throttles stream of offsets depend on power of scroll. And all i need is stop scrolling in particular offset  - 500px from top, without overscrolling.

Do I miss something, or should i use another workaround for my task?

Thank you for reply!

**Comment by pixel-0ne** [comment]


**Reply by dngraphisme** [reply]
Hello, i want a solution for jQuery please :(

Thanks for your help and your library

**Reply by dolphin-wood** [reply]
I don't really understand the issue here: the above code was written in plain JS and it should work no matter what library/framework you are using. If you want a solution in es5, you can either refer to https://github.com/idiotWu/smooth-scrollbar/issues/104#issuecomment-335286699 or paste the above code into babel transpiler.

**Comment by Mak-Pro** [comment]
I changed the code of the plugin itself a bit. I know this is not entirely correct. But no other idea came to mind.
You need to unminify the plugin code. And add a check for the presence of a class, which can be added when you need to pause the scroll.
Just find this pieces of code, and modify them like below

This is for desktop:
function st(t) {

            F(t)(t.options.delegateTo || t.containerEl, "onwheel" in window || document.implementation.hasFeature("Events.wheel", "3.0") ? "wheel" : "mousewheel", function (e) {

                if(!t.contentEl.classList.contains('**fixed**')) {
                    var n = (function (t) {
                        if ("deltaX" in t) {
                            var e = pt(t.deltaMode);
                            return { x: (t.deltaX / ft.STANDARD) * e, y: (t.deltaY / ft.STANDARD) * e };
                        }
                        return "wheelDeltaX" in t ? { x: t.wheelDeltaX / ft.OTHERS, y: t.wheelDeltaY / ft.OTHERS } : { x: 0, y: t.wheelDelta / ft.OTHERS };
                    })(e),
                    r = n.x,
                    o = n.y;
                    t.addTransformableMomentum(r, o, e, function (t) {
                        t && e.preventDefault();
                    });
                }
            });


        }

This is for mobile:

i(r, "touchmove", function (e) {
                if(!t.contentEl.classList.contains('**fixed**')) {
                    if (!et || et === t) {
                        o.update(e);
                        var n = o.getDelta(),
                        r = n.x,
                        i = n.y;
                        t.addTransformableMomentum(r, i, e, function (n) {
                            n && e.cancelable && (e.preventDefault(), (et = t));
                        });
                    }
                }
                
            }),

**Comment by natainditama** [comment]
thank you

**Reply by anku598** [reply]
Ho can i do this in the nuxtjs ? i want the same functionality when my nav is open and restore it when it false
